project-koku/koku
# [COST-3991] daily-archive-idea

## Jira Ticket

[COST-3991](https://issues.redhat.com/browse/COST-3991)

## Description

This change will add daily archive flows to Azure and AWS download processing. This will enable us to break down our download processing to only reprocess a few days rather than a full month every time a download is triggered. This also allows us to more accurately pass on start/end dates to summary tasks for the processed files.

Currently this is going to need some semi manual S3 file purging/reprocessing triggered when its released.

## Testing

1. Checkout Branch
2. Restart Koku
3. Load test data
4. See everything ingests and functions like normal

## Notes
Requires a parquet compactor change https://github.com/project-koku/parquet-compactor/pull/30

Advanced testing:
AWS!!
1. Create provider with multiple months of data that you can generate new nise reports
2. See inital ingest process daily files for the entire two months
3. Generate new nise data for current month and hit the download endpoint
4. Monitor the logs and see we only convert to parquet/summarise the last 3 days of data since the last update timestamp on that provider (today in our case)
5. Now generate NEW data for previous month
6. Hit download
7. See we also only convert to parquet/summarise the last 4 days of the previous month.
8. Now generate more previous month data WITH a finalized_id
9. Trigger download
10. See we now clear out all parquet files (this is in case some got compacted) then we process the entire previous month daily files again for the final time.


Azure -- There is no finalization in Azure
1. Create provider with multiple months of data that you can generate new nise reports
2. See inital ingest process daily files for the entire two months
3. Generate new nise data for current month and hit the download endpoint
4. Monitor the logs and see we only convert to parquet/summarise the last 4 days of data since the last update timestamp on that provider (today in our case)
5. Now generate NEW data for previous month
6. Hit download endpoint
7. See we now clear out all parquet files (this is in case some got compacted) then we process the entire previous month daily files again for the final time.

Main to This PR testing!

AWS/Azure
1. Create Provider/ingest data
2. Check Minio files have a single CSV/Parquet file for the full month of data
3. Check counts from DB [0]
4. checkout this branch
5. run migrations
6. create a new payload for your provider
7. Hit download

YES!! No need to hand held the release of this one any more!

[0] - select usage_start, sum(unblended_cost) from reporting_awscostentrylineitem_daily_summary group by usage_start;
